image/registry_handler.go

```diff
@@ -0,0 +1,99 @@
+package image
+
+import (
+	"strings"
+
+	"github.com/buildpacks/imgutil/remote"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/pkg/errors"
+
+	"github.com/buildpacks/lifecycle/cmd"
+)
+
+// RegistryHandler takes care of the registry settings and checks
+//
+//go:generate mockgen -package testmock -destination testmock/registry_handler.go github.com/buildpacks/lifecycle RegistryHandler
+type RegistryHandler interface {
+	EnsureReadAccess(imageRefs ...string) error
+	EnsureWriteAccess(imageRefs ...string) error
+}
+
+// DefaultRegistryHandler is the struct that implements the RegistryHandler methods
+type DefaultRegistryHandler struct {
+	keychain         authn.Keychain
+	insecureRegistry []string
+}
+
+// NewRegistryHandler creates a new DefaultRegistryHandler
+func NewRegistryHandler(keychain authn.Keychain, insecureRegistries []string) *DefaultRegistryHandler {
+	return &DefaultRegistryHandler{
+		keychain:         keychain,
+		insecureRegistry: insecureRegistries,
+	}
+}
+
+// EnsureReadAccess ensures that we can read from the registry
+func (rv *DefaultRegistryHandler) EnsureReadAccess(imageRefs ...string) error {
+	for _, imageRef := range imageRefs {
+		if err := verifyReadAccess(imageRef, rv.keychain, GetInsecureOptions(rv.insecureRegistry, imageRef)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// EnsureWriteAccess ensures that we can write to the registry
+func (rv *DefaultRegistryHandler) EnsureWriteAccess(imageRefs ...string) error {
+	for _, imageRef := range imageRefs {
+		if err := verifyReadWriteAccess(imageRef, rv.keychain, GetInsecureOptions(rv.insecureRegistry, imageRef)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GetInsecureOptions returns a list of WithRegistrySetting imageOptions matching the specified imageRef prefix
+/*
+TODO: This is a temporary solution in order to get insecure registries in other components too
+TODO: Ideally we should fix the `imgutil.options` struct visibility in order to mock and test the `remote.WithRegistrySetting`
+TODO: function correctly and use the RegistryHandler everywhere it is needed.
+*/
+func GetInsecureOptions(insecureRegistries []string, imageRef string) []remote.ImageOption {
+	var opts []remote.ImageOption
+	if len(insecureRegistries) > 0 {
+		for _, insecureRegistry := range insecureRegistries {
+			if strings.HasPrefix(imageRef, insecureRegistry) {
```

Comment:
Uhm, the insecure-registry feature disables any certificate checks and it should be used only in certain occasion (like testing and towards a registry in a secure network).

This check has been done to verify that one of the `--insecure-registry=<registry.address>` is the same as the one provided in the refImage (like `<registry.address>/app-name`)

So personally I would keep that check just to avoid any human error.
Does it make sense?

Reply:
Does the `insecureRegistry` on `remote.WithRegistrySetting(insecureRegistry, true)` not limit this to the registry internally?

Reply:
Uhm not that I'm aware of: https://github.com/buildpacks/imgutil/blob/main/remote/options.go#L101C6-L101C25

But I guess as a next improvement we can integrate this check inside the `NewImage` function if we want to

Reply:
I tried spinning up two http registries and couldn't find a combination of `remote` options to make it work as I expected. In this example I have 2 insecure registries and I would like to be able to use them both in creation of a single image.

Note: I'm using imgutil here and not lifecycle, as I wanted to see how the registry primitives work.

```
	baseImageName := "myregistry.domain.com:6443/test/test:latest"
	destinationImageName := "myregistry.domain.com:7443/test/test:latest"

	remoteImage, err := remote.NewImage(destinationImageName,
		authn.DefaultKeychain,
		remote.FromBaseImage(baseImageName),
		remote.WithRegistrySetting("myregistry.domain.com:6443", true),
		remote.WithRegistrySetting("myregistry.domain.com:7443", true),
	)
```

The image registry setting for the base repo is not applied. I always get

```
connect to repo store "myregistry.domain.com:6443/test/test:latest": Get "https://myregistry.domain.com:6443/v2/": http: server gave HTTP response to HTTPS client
```

I can get a single registry setting to work - but not both. Maybe this is expected and I'm doing something wrong though.

Reply:
Uhm, according to what I tested on lifecycle:
1. check the registry with https failing
2. try with http

So I guess it just use the first one and stop whenever it succeed.
Did you try to make the first failing but the second to succeed?

btw thank you for these tests! I didn't spawn up an insecure-registry with imgutil, (if you have any doc I can follow to reproduce those steps would be great!), I just relied on lifecycle acceptance tests

Reply:
I followed the instructions [here](https://docs.docker.com/registry/insecure/#use-self-signed-certificates) for spinning up a registry with a self-signed cert. Create the cert, update your host file, and then I spun up a few in different terminals like this.

```
docker run -it -v "$(pwd)"/certs:/certs \
  -e REGISTRY_HTTP_ADDR=0.0.0.0:443 \
  -e REGISTRY_HTTP_TLS_CERTIFICATE=/certs/domain.crt \
  -e REGISTRY_HTTP_TLS_KEY=/certs/domain.key \
  -p 6443:443 \
  registry:2

docker run -it -v "$(pwd)"/certs:/certs \
  -e REGISTRY_HTTP_ADDR=0.0.0.0:443 \
  -e REGISTRY_HTTP_TLS_CERTIFICATE=/certs/domain.crt \
  -e REGISTRY_HTTP_TLS_KEY=/certs/domain.key \
  -p 7443:443 \
  registry:2


docker run -it  -e REGISTRY_HTTP_ADDR=0.0.0.0:8080 \
  -p 8080:8080 \
  registry:2

docker run -it  -e REGISTRY_HTTP_ADDR=0.0.0.0:8080 \
  -p 9080:8080 \
  registry:2
```

This way I had a combination of HTTP/TLS registries to play around with as target/base registries. I didn't do a full test matrix or anything at this point. We can use `crane` to compare behavior as well when it makes sense. If you aren't diving back into `imgutil` yourself, let me know and I will try to make some time to make changes and add tests there soon™️ . I think we have to get `imgutil` in a good spot before we land any lifecycle changes in case the interfaces need to change.

Reply:
Great thanks!
Indeed I'm still trying to understand how imgutil works under the hood and tested that parts wasn't really easy so your help would be really appreciated and yes I agree with you that we should have always imgutil in a good shape being a very important library for lifecycle.
If you have time to adjust/test it that would be awesome, I can learn a bit more about its internals 😄

Reply:
Would this be fixed by https://github.com/buildpacks/imgutil/pull/220?

Reply:
I'm not sure. I still think we need to always `remote.WithRegistrySetting(insecureRegistry, true)` for ALL the insecure registries all the time. The contains check doesn't make sense to me, but I may be missing something.

In the case of creating a new image in an insecure registry, but using a _different_ insecure registry as the base, could we do something like this?

```golang
appImage, err := remote.NewImage(
		e.OutputImageRef,
		e.keychain,
                image.GetInsecureOptions(r.InsecureRegistries), // sets up ALL the insecure registries that may be used
                remote.FromBaseImage(r.RunImageRef),		
	)
```

If we make it so that the caller has to pass the image into `image.GetInsecureOptions` for each image, it means we it would look like this.

```golang
appImage, err := remote.NewImage(
		e.OutputImageRef,
		e.keychain,
                image.GetInsecureOptions(r.InsecureRegistries, e.RunImageRef),
                image.GetInsecureOptions(r.InsecureRegistries, e.OutputImageRef),
		opts...,
	)
```

That seems harder to do for no benefit. I don't think there is any harm in always having the map with all the possible insecure registries..even if NONE of the images reference it. It's like having the `insecureRegistries` value in a docker config.

Reply:
It makes sense to me, I just modified the `GetInsecureOptions` implementation to accept all insecure-registries.
Let me know what do you think 😄